File: 713-subarray-product-less-than-k/subarray-product-less-than-k.java
class Solution 
{
    public int numSubarrayProductLessThanK(int[] nums, int k)
    {
        if(k==0)
        return 0;
        int n=nums.length;
        int l=0,r=0;
        int prod=1;
        int c=0;
        while(r<n)
        {
            prod*=nums[r];
            r++;
            while(prod>=k && l<r)
            {
                prod/=nums[l];
                l++;
            }
            c+=(r-l);
        }
        return c;
    }
}

File: 713-subarray-product-less-than-k/subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        l,r = 0,0
        ans = 1
        res = 0
        n = len(nums)
        while r<n:
            ans *= nums[r]
            while l<=r and ans>=k:
                ans //= nums[l]
                l+=1
            res += (r-l+1)
            r+=1
        return res